File: 22.Automated-Testing-Unit-Integration-and-E2E/code/typescript/dependency_injection_for_testability_4.ts
interface Order { total: number; }
interface OrderRepo {
    save(o: Order): Promise<void>;
}

class OrderService {
    constructor(private repo: OrderRepo) {}

    async place(o: Order): Promise<void> {
        if (o.total <= 0) throw new Error("invalid total");
        await this.repo.save(o);
    }
}

describe('OrderService', () => {
    it('rejects invalid total', async () => {
        const fakeRepo: OrderRepo = {
            saved: [],
            save: async function(o: Order) { this.saved.push(o); }
        } as any;

        const svc = new OrderService(fakeRepo);
        await expect(svc.place({ total: 0 })).rejects.toThrow("invalid total");
    });
});

File: 22.Automated-Testing-Unit-Integration-and-E2E/code/rust/dependency_injection_for_testability_4.rs
struct Order { total: i32 }
trait OrderRepo { fn save(&mut self, o: Order) -> Result<(), ()>; }

struct OrderService<R: OrderRepo> { repo: R }

impl<R: OrderRepo> OrderService<R> {
    fn place(&mut self, o: Order) -> Result<(), ()> {
        if o.total <= 0 { return Err(()); }
        self.repo.save(o)
    }
}

struct FakeRepo { saved: Vec<Order> }
impl OrderRepo for FakeRepo {
    fn save(&mut self, o: Order) -> Result<(), ()> {
        self.saved.push(o); Ok(())
    }
}

#[test]
fn test_place_rejects_invalid_total() {
    let mut svc = OrderService { repo: FakeRepo { saved: vec![] } };
    assert!(svc.place(Order { total: 0 }).is_err());
}
